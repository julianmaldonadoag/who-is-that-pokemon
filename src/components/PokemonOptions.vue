<template>
  <div class="options-container">
    <ul>
      <li 
        v-for="pokemon in pokemons" 
        :key="pokemon.id"
        :class="{ 
          'disable': disableBtn, 
          'correct-answer': correctAnswer(pokemon.id),
          'incorrect-answer': incorrectAnswer(pokemon.id)
        }"
        @click="$emit('chosenPokemon', pokemon.id); chosenPokemon(pokemon);"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  data() {
    return {
      disableBtn: false,
      pokemonSelected: null
    }
  },
  props: {
    pokemons: {
      type: Array,
      required: true
    },
    correctPokemon: {
      type: Object,
      required: true
    }
  },
  methods: {
    chosenPokemon(pokemon) {
      this.disableBtn = true
      this.pokemonSelected = pokemon
    },
    correctAnswer(pokemonId) {
      return this.disableBtn && pokemonId === this.pokemonSelected.id && pokemonId === this.correctPokemon.id
    },
    incorrectAnswer(pokemonId) {
      return this.disableBtn && pokemonId === this.pokemonSelected.id && pokemonId !== this.correctPokemon.id
    }
  }
}
</script>

<style scoped>
.options-container {
  display: flex;
  justify-content: center;
}
ul {
  list-style-type: none;
  padding-left: 0;
}
li {
  background-color: #0d6efd;
  color: #fff;
  border-radius: 5px;
  cursor: pointer;
  margin-bottom: 10px;
  padding: 10px 0;
  width: 250px;
}
li:hover {
  background-color: #084298;
}
.disable {
  pointer-events: none;
}
.correct-answer {
  background-color: #198754;
}
.incorrect-answer {
  background-color: #dc3545;
}
</style>