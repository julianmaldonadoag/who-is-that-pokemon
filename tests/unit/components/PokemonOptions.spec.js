import { shallowMount } from "@vue/test-utils"
import PokemonOptions from "@/components/PokemonOptions.vue"
import { pokemons } from "../mocks/pokemons.mock"

describe("PokemonOptions component", () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PokemonOptions, {
      props: {
        pokemons,
      },
    })
  })

  test("should match with snapshot", () => {
    // console.log(wrapper.html())

    expect(wrapper.html()).toMatchSnapshot()

    // the value of toMatchInlineSnapshot was autogenerated when test was run
    // expect(wrapper.html()).toMatchInlineSnapshot(`
    //   <div class="options-container">
    //     <ul>
    //       <li>bulbasaur</li>
    //       <li>ivysaur</li>
    //       <li>venusaur</li>
    //       <li>charmander</li>
    //     </ul>
    //   </div>
    // `)
  })

  test('should show 4 pokemon names correctly', () => {
    const liTags = wrapper.findAll('li')

    expect(liTags.length).toBe(4)
    expect(liTags[0].text()).toBe('bulbasaur')
    expect(liTags[1].text()).toBe('ivysaur')
    expect(liTags[2].text()).toBe('venusaur')
    expect(liTags[3].text()).toBe('charmander')
  })

  test('should emit "chosenPokemon" with its respective params', () => {
    const [li1, li2, li3, li4] = wrapper.findAll('li')

    li1.trigger('click')
    li2.trigger('click')
    li3.trigger('click')
    li4.trigger('click')

    console.log(wrapper.emitted('chosenPokemon'))

    expect(wrapper.emitted('chosenPokemon').length).toBe(4)
    expect(wrapper.emitted('chosenPokemon')[0]).toEqual([1])
    expect(wrapper.emitted('chosenPokemon')[1]).toEqual([2])
    expect(wrapper.emitted('chosenPokemon')[2]).toEqual([3])
    expect(wrapper.emitted('chosenPokemon')[3]).toEqual([4])
  })
})
