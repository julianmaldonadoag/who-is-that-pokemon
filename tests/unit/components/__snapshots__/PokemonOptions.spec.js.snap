// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions component should match with snapshot 1`] = `
<div class="options-container">
  <ul>
    <li class="">bulbasaur</li>
    <li class="">ivysaur</li>
    <li class="">venusaur</li>
    <li class="">charmander</li>
  </ul>
</div>
`;
